iwyu.cc

```diff
@@ -2594,8 +2583,17 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     vector<const UsingDecl*> using_decls
         = FindInMultiMap(visitor_state_->using_declarations, decl);
     for (Each<const UsingDecl*> it(&using_decls); !it.AtEnd(); ++it) {
-      if (!(*it)->getDeclContext()->Encloses(using_context))
-        continue;
+      const DeclContext* using_decl_context = (*it)->getDeclContext();
+      if (!using_decl_context->Encloses(use_context)) {
```

Comment:
Thanks for the explanation, this makes more sense to me now.

There's one branch where this seems broken; if Encloses returns false and the decl-kind is a function, we'll always register it, whether it's the *right* function (i.e. one enclosing the using-decl) or not.

It seems like this could cause problems, but I'm not sure.

I think if Encloses fails you could just check if the using-decl's parent is the same as the decl's parent, and if either of them is a FunctionDecl. That seems like a naive implementation of Encloses :-). I'll ask the Clang devs if Encloses is broken in this regard, it could be that we could get it fixed.